src/components/document/annotation-layer.tsx

```diff
@@ -397,6 +408,64 @@ export const AnnotationLayer = observer(function AnnotationLayer({
   const hidden = !persistentUI.showAnnotations;
   const classes = classNames("annotation-layer",
     { editing, hidden, 'show-buttons': showButtons, 'show-handles': showDragHandles });
+
+  const renderAnnotationButtons = (params: {
+    tile: ITileModel;
+    rowId: string;
+    tileId: string;
+    parentOffsetParams?: { rowId: string; tileId: string };
+  }) => {
+    const { tile, rowId, tileId, parentOffsetParams } = params;
+
+    return tile.content.annotatableObjects.map(({ objectId, objectType }) => (
+      <AnnotationButton
+        key={`${tile.id}-${objectId}-button`}
+        getObjectBoundingBox={getObjectBoundingBox}
+        getTileOffset={() => getTileOffset(rowId, tileId, parentOffsetParams)}
+        objectId={objectId}
+        objectType={objectType}
+        onClick={handleAnnotationButtonClick}
+        sourceObjectId={sourceObjectId}
+        sourceTileId={sourceTileId}
+        tileId={tile.id}
+      />
+    ));
+  };
+
+  const collectButtonsForRow = (
+    rowId: string,
+    parentOffsetParams?: { rowId: string; tileId: string },
+    tileContent?: DocumentContentModelType | QuestionContentModelType
+  ): JSX.Element[] => {
+
+    const docContent = content;
+    const row = tileContent ? tileContent?.rowMap.get(rowId) : docContent?.rowMap.get(rowId);
+
+    if (!row) return [];
+    return row.tiles.flatMap((tileInfo) => {
+      const tile = docContent?.tileMap?.get(tileInfo.tileId);
+      if (!tile) return [];
+
+      // Question tile: dive into its own rows
+      if (tile.content.type === kQuestionTileType) {
+        const qTileContent = tile.content as QuestionContentModelType;
+        const newParentOffset = { rowId, tileId: tile.id };
+        return qTileContent.rowOrder.flatMap((nestedRowId) => {
+          return collectButtonsForRow(nestedRowId, newParentOffset, qTileContent);
+        });
+      }
```

Comment:
I think this can be simplified a bit, and ideally avoid direct references to Question Tile (in case other container types are added in the future):

`DocumentContentModelType | QuestionContentModelType` --> can use the common parent, `RowListType`.

`tileContent ? tileContent?.rowMap.get(rowId) : docContent?.rowMap.get(rowId)` --> simplify to `docContent.getRowRecursive(rowId)`.

`if (tile.content.type === kQuestionTileType)` --> `const qTileContent=tile.content; if (isRowListContainer(qTileContent))` should avoid the need to cast it.

Reply:
Oh nice - I didn't realize I could use getRowRecursive and isRowListContainer. That definitely makes it simpler!